employees/views.py

```diff
@@ -204,14 +230,21 @@ def post(self, request: HttpRequest, pk: int) -> Union[Response, HttpResponseRed
         return redirect("custom-report-list")
 
 
-def delete_report(_request: HttpRequest, pk: int) -> HttpResponseRedirectBase:
-    report = get_object_or_404(Report, pk=pk)
-    report.delete()
-    logger.debug(f"Report with id: {pk} has been deleted")
-    return redirect("custom-report-list")
+@method_decorator(login_required, name="dispatch")
+@method_decorator(
+    check_permissions(allowed_user_types=[CustomUser.UserType.EMPLOYEE.name, CustomUser.UserType.ADMIN.name]),
+    name="dispatch",
+)
+class ReportDeleteView(UserIsAuthorOfCurrentReportMixin, DeleteView):
```

Comment:
Nothing I can think of. Consider this example:

There is a view class that has `get_queryset()` method with some complex filtering. If you would like to do this in decorator too, you would have to call and evaluate `get_queryset()` method of decorated class inside decorator (which sounds ugly already) to figure out if returned object match current user/role.

If this would be added in the view code in form of `if`, it would require overwriting method in each view in which we would like to add it. Sounds like a lot of work.

I think the mixin approach and limiting queryset is the cleanest and most optimized way to do this. At least of all those which I have thought of, but maybe there is a better way I haven't thought of.

The part I don't like with this solution is that it returns 404 instead of 403, but I think its fair trade-off.

> if not, we redirect him to login page (?)

Yes, exactly this is done in this decorator.

Reply:
@rwrzesien Ok, I'am convinced, thanks for the explanation :)